src/dawn/IIR/DoMethod.cpp

```diff
@@ -85,6 +90,91 @@ const std::shared_ptr<DependencyGraphAccesses>& DoMethod::getDependencyGraph() c
   return derivedInfo_.dependencyGraph_;
 }
 
+void DoMethod::DerivedInfo::clear() {
+  fields_.clear();
+}
+
+void DoMethod::clearDerivedInfo() { derivedInfo_.clear(); }
+
+void DoMethod::updateLevel() {
+
+  // Compute the fields and their intended usage. Fields can be in one of three states: `Output`,
+  // `InputOutput` or `Input` which implements the following state machine:
+  //
+  //    +-------+                               +--------+
+  //    | Input |                               | Output |
+  //    +-------+                               +--------+
+  //        |                                       |
+  //        |            +-------------+            |
+  //        +----------> | InputOutput | <----------+
+  //                     +-------------+
+  //
+  std::unordered_map<int, Field> inputOutputFields;
+  std::unordered_map<int, Field> inputFields;
+  std::unordered_map<int, Field> outputFields;
+
+  for(const auto& statementAccessesPair : children_) {
+    const auto& access = statementAccessesPair->getAccesses();
+    DAWN_ASSERT(access);
+
+    for(const auto& accessPair : access->getWriteAccesses()) {
+      int AccessID = accessPair.first;
+      Extents const& extents = accessPair.second;
+
+      // Does this AccessID correspond to a field access?
+      if(!stencilInstantiation_.isField(AccessID)) {
+        continue;
+      }
+      AccessUtils::recordWriteAccess(inputOutputFields, inputFields, outputFields, AccessID,
+                                     extents, getInterval());
+    }
+
+    for(const auto& accessPair : access->getReadAccesses()) {
+      int AccessID = accessPair.first;
+      Extents const& extents = accessPair.second;
+
+      // Does this AccessID correspond to a field access?
+      if(!stencilInstantiation_.isField(AccessID)) {
+        continue;
+      }
+
+      AccessUtils::recordReadAccess(inputOutputFields, inputFields, outputFields, AccessID, extents,
+                                    getInterval());
+    }
+  }
+
+  // Merge inputFields, outputFields and fields
+  derivedInfo_.fields_.insert(outputFields.begin(), outputFields.end());
+  derivedInfo_.fields_.insert(inputOutputFields.begin(), inputOutputFields.end());
+  derivedInfo_.fields_.insert(inputFields.begin(), inputFields.end());
+
+  if(derivedInfo_.fields_.empty()) {
+    DAWN_LOG(WARNING) << "no fields referenced in stage";
+    return;
+  }
+
+  // Compute the extents of each field by accumulating the extents of each access to field in the
+  // stage
+  for(const auto& statementAccessesPair : iterateIIROver<StatementAccessesPair>(*this)) {
+    const auto& access = statementAccessesPair->getAccesses();
+
+    // first => AccessID, second => Extent
```

Comment:
i like this, should we start doing this everywhere? or should the names reflect this?

Reply:
yes, make sense. We could try to do both, for readability